### PR TITLE
SLI-861 Include security hotspots in analysis

### PR DESCRIPTION
**Pull request by damien-urruty-sonarsource**

### CODE REVIEW

File: src/main/java/org/sonarlint/intellij/actions/SonarLintToolWindow.java [hunk 1]
```diff
@@ -59,19 +62,32 @@ public SonarLintToolWindow(Project project) {
   /**
    * Must run in EDT
    */
-  public void openAnalysisResults() {
+  public void openReportTab(AnalysisResult analysisResult) {
+    this.<ReportPanel>withTab(SonarLintToolWindowFactory.REPORT_TAB_TITLE, panel -> panel.updateIssues(analysisResult));
+  }
+  public void clearReportTab() {
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/actions/SonarLintToolWindow.java:68]:
```suggestion
  }
  
  public void clearReportTab() {
```

File: src/main/java/org/sonarlint/intellij/actions/ClearReportAction.java [hunk 2]
```diff
@@ -22,12 +22,11 @@
...
-import org.jetbrains.annotations.Nullable;
-import org.sonarlint.intellij.common.util.SonarLintUtils;
-import org.sonarlint.intellij.issue.IssueStore;
 
-public class SonarClearAnalysisResultsAction extends AnAction {
-  public SonarClearAnalysisResultsAction(@Nullable String text, @Nullable String description, @Nullable Icon icon) {
+import static org.sonarlint.intellij.common.util.SonarLintUtils.getService;
+
+public class ClearReportAction extends AnAction {
+  public ClearReportAction(String text, String description, Icon icon) {
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/actions/ClearReportAction.java:29]: We lost information of `@Nullable` on constructor parameters

File: src/main/java/org/sonarlint/intellij/core/SonarLintEngineFactory.java [hunk 3]
```diff
@@ -53,6 +53,7 @@ ConnectedSonarLintEngine createEngine(String connectionId, boolean isSonarCloud)
     var configBuilder = isSonarCloud ? ConnectedGlobalConfiguration.sonarCloudBuilder() : ConnectedGlobalConfiguration.sonarQubeBuilder();
     configBuilder
       .addEnabledLanguages(EmbeddedPlugins.getEnabledLanguagesInConnectedMode().toArray(new Language[0]))
+      .enableHotspots()
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/core/SonarLintEngineFactory.java:56]: Should be enabled only in the case of `sonarQubeBuilder`?

File: src/main/java/org/sonarlint/intellij/finding/LiveFinding.java [hunk 4]
```diff
@@ -173,7 +167,7 @@ public IssueSeverity getUserSeverity() {
 
   @Override
   public RuleType getType() {
-    return type;
+    return RuleType.SECURITY_HOTSPOT;
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/finding/LiveFinding.java:170]: Why do we force `RuleType` to be a `SECURITY_HOTSPOT`? `getType()` is also called for issues

File: src/main/java/org/sonarlint/intellij/finding/RawIssueAdapter.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+          var locInputFile = loc.getInputFile();
+          if (textRange != null && locInputFile != null) {
+            var locPsiFile = toPsiFile(project, locInputFile.getClientObject());
+            var range = matcher.match(locPsiFile, textRange);
+            matchedLocations.add(resolvedLocation(locPsiFile.getVirtualFile(), range, loc.getMessage(), null));
+          }
+        } catch (TextRangeMatcher.NoMatchException e) {
+          // File content is likely to have changed during the analysis, should be fixed in next analysis
+          SonarLintConsole.get(project)
+            .debug("Failed to find secondary location of issue for file: '" + psiFile.getName() + "'. The location won't be displayed - " + e.getMessage());
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/finding/RawIssueAdapter.java:95]:
```suggestion
            .debug("Failed to find secondary location of finding for file: '" + psiFile.getName() + "'. The location won't be displayed - " + e.getMessage());
```

File: src/main/java/org/sonarlint/intellij/core/SonarLintEngineFactory.java (hunk 3 above)
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/core/SonarLintEngineFactory.java:56]: In the case of SonarCloud we should not enable hotspots for now

File: src/main/java/org/sonarlint/intellij/finding/persistence/FindingsManager.java [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        } catch (Exception e) {
+          SonarLintConsole.get(myProject).error("Cannot flush issues", e);
+        }
+      }
+    });
+  }
+
+  public void clearAllIssuesForAllFiles() {
+    liveIssueCache.clear();
+    myProject.getMessageBus().syncPublisher(IssueStoreListener.SONARLINT_ISSUE_STORE_TOPIC).allChanged();
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/finding/persistence/FindingsManager.java:89]: What about renaming IssueStoreListener to FindingStoreListener?

File: src/main/java/org/sonarlint/intellij/finding/persistence/FindingsManager.java [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+  }
+
+  public void removeResolvedFindings(VirtualFile file, List<LiveIssue> issues, List<LiveSecurityHotspot> securityHotspots) {
+    liveIssueCache.removeFindings(file, issues);
+    liveSecurityHotspotCache.removeFindings(file, securityHotspots);
+    notifyFindingsChangedForFile(file);
+  }
+
+  public boolean neverAnalyzed(VirtualFile file) {
+    return liveIssueCache.getLive(file) == null;
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/finding/persistence/FindingsManager.java:111]: I think we should also check `&& liveSecurityHotspotCache.getLive(file) == null`

File: src/main/java/org/sonarlint/intellij/finding/tracking/Tracker.java [hunk 8]
```diff
@@ -373,7 +373,7 @@ private static class ServerIssueSearchKey implements SearchKey {
     private final String serverIssueKey;
 
     ServerIssueSearchKey(Trackable trackable) {
-      serverIssueKey = trackable.getServerIssueKey();
+      serverIssueKey = trackable.getServerFindingKey();
```
Comment (nquinquenel) [inline src/main/java/org/sonarlint/intellij/finding/tracking/Tracker.java:376]: I don't think `serverIssueKey` makes sense if we're using Finding now, what about renaming inner class `ServerIssueSearchKey`  and enum `ServerIssueSearchKeyFactory`

File: src/main/java/org/sonarlint/intellij/core/SonarLintEngineFactory.java (hunk 3 above)
Comment (damien-urruty-sonarsource) [reply src/main/java/org/sonarlint/intellij/core/SonarLintEngineFactory.java:56]: This `enableHotspots` flag means "enable if the server permits it", so anyway we check if it's SC [a bit later](https://github.com/SonarSource/sonarlint-core/blob/master/server-api/src/main/java/org/sonarsource/sonarlint/core/serverapi/hotspot/HotspotApi.java#L59) down the calls. I'm tempted to leave it like this, the day we can get hotspots from SC we don't have to change SLI

File: src/main/java/org/sonarlint/intellij/finding/persistence/FindingsManager.java (hunk 6 above)
Comment (damien-urruty-sonarsource) [reply src/main/java/org/sonarlint/intellij/finding/persistence/FindingsManager.java:89]: Yes. And I hope to kill this interface in a next refactoring :)

File: src/main/java/org/sonarlint/intellij/finding/persistence/FindingsManager.java (hunk 7 above)
Comment (damien-urruty-sonarsource) [reply src/main/java/org/sonarlint/intellij/finding/persistence/FindingsManager.java:111]: Good catch! And I see we have a `neverAnalyzed` and `wasAnalyzed`, I will get rid of one of them

Comment (damien-urruty-sonarsource) [reply src/main/java/org/sonarlint/intellij/finding/persistence/FindingsManager.java:111]: In fact the meaning of both methods is slightly different so I will keep them both but fix the issue and rename one

File: src/main/java/org/sonarlint/intellij/finding/tracking/Tracker.java (hunk 8 above)
Comment (damien-urruty-sonarsource) [reply src/main/java/org/sonarlint/intellij/finding/tracking/Tracker.java:376]: Yes

**Review by henryju — APPROVED** [review]
Didn't look in detail, but overall LGTM. I suggest merging in the common branch to not block the rest of the sprint

**Review by nquinquenel — APPROVED** [review]

**Pull request merged**